client/proto/state_snapshot/v1/state_snapshot.proto

```diff
@@ -0,0 +1,159 @@
+  //                               [State Snapshot at evgN]
+  //                                          |
+  //                                          v
+  // ------------------                                            ---------------
+  // | Concord Client |  -- evg1, evg2, ..., evgN, evgN+1, ... ->  | Application |
+  // ------------------                                            ---------------
+  // -----------------------------------------------------------------------------
+  // where `evgN` is the Nth event group streamed from Concord Client to the Application.
+  // After receiving a state snapshot at `evgN`, an Application can continue operation by
+  // consuming event groups from `evgN+1` onwards.
+  //
+  // Concord Client ensures that there is always at least one reasonably recent
+  // state snapshot that can be served. The intention is that this state snapshot
+  // can be used to initialize the Application after the blockchain has been pruned and
+  // some set of historical events are no longer available via the `EventService`.
+  //
+  // See `event.proto` for more information about events, event group IDs and
+  // `EventService`.
+  //
+  // Note: The StateSnapshot service only supports state snapshots based on event
+  // group IDs and not legacy events that are based on block IDs.
+  uint64 event_group_id = 2;
+
+  // Mandatory field.
+  //
+  // An estimate (with reasonable accuracy) of the count of key-values contained
+  // in the state snapshot. Please note that this is an estimation and *not* the
+  // actual count.
+  uint64 key_value_count_estimate = 3;
+}
+
+message StreamSnapshotRequest {  
+  // Mandatory field.
+  //
+  // The ID of the state snapshot to be streamed.
+  uint64 snapshot_id = 1;
+
+  // Optional field.
+  //
+  // If set, start streaming from `last_received_key` onwards, excluding
+  // `last_received_key`.
+  // If `last_received_key` is not part of the state snapshot, an INVALID_ARGUMENT
+  // error is returned.
+  //
+  // If not set, start streaming from the first key-value in the state snapshot.
+  //
+  // The empty bytestring is a valid key.
+  //
+  // Key-values are streamed with lexicographic order on keys.
+  optional bytes last_received_key = 2;
+}
+
+message KeyValuePair {
+  // Mandatory field.
+  //
+  // The key.
+  bytes key = 1;
+
+  // Mandatory field.
+  //
+  // The value.
+  bytes value = 2;
+}
+
+message StreamSnapshotResponse {
+  // Mandatory field.
+  KeyValuePair key_value = 1;
+}
+
+message ReadAsOfRequest {
+  // Mandatory field.
+  //
+  // The ID of the state snapshot as of which to read the values.
+  uint64 snapshot_id = 1;
+
+  // List of keys for which the values should be returned.
+  repeated bytes keys = 2;
+}
+
+message OptionalValue {
+  optional bytes value = 1;
+}
+
+message ReadAsOfResponse {
+  // The `values` list contains entries for all requested keys
+  // in `ReadAsOfRequest.keys`, i.e. it is the exact same length.
```

Comment:
We need to specify here the order of the values as the response list doesn't contain the requested keys.

Reply:
Yes, you are right. I thought I've done it, but I haven't. Will do so.